django/template/defaulttags.py

```diff
@@ -1570,3 +1572,94 @@ def do_with(parser, token):
     nodelist = parser.parse(("endwith",))
     parser.delete_first_token()
     return WithNode(None, None, nodelist, extra_context=extra_context)
+
+
+class DefinePartialNode(Node):
+    def __init__(self, partial_name, inline, nodelist):
+        self.partial_name = partial_name
+        self.inline = inline
+        self.nodelist = nodelist
+
+    def render(self, context):
+        return self.nodelist.render(context) if self.inline else ""
+
+
+class RenderPartialNode(Node):
+    def __init__(self, partial_name, partial_mapping):
+        # Defer lookup of nodelist to runtime.
+        self.partial_name = partial_name
+        self.partial_mapping = partial_mapping
+
+    def render(self, context):
+        return self.partial_mapping[self.partial_name].render(context)
+
+
+@register.tag(name="partialdef")
+def partialdef_func(parser, token):
+    """
+    Declare a partial that can be used later in the template.
+
+    Usage::
+
+        {% partialdef partial_name %}
+        Partial content goes here
+        {% endpartialdef %}
+
+    Stores the nodelist in the context under the key "partial_contents" and can
+    be retrieved using the {% partial %} tag.
+
+    The optional ``inline`` argument will render the contents of the partial
+    where it is defined.
+    """
+    tokens = token.split_contents()
+
+    # Check the number of tokens before trying to assign them via indexes.
+    if (tokens_len := len(tokens)) not in (2, 3):
+        name = token.contents.split()[0]
+        raise TemplateSyntaxError(f"{name} tag requires 2-3 arguments")
+
+    partial_name = tokens[1]
+    if tokens_len > 2:
+        inline = tokens[2]
+        if inline != "inline":
+            raise TemplateSyntaxError(
+                "The 'inline' argument does not have any parameters; "
+                "either use 'inline' or remove it completely."
+            )
+    else:
+        inline = False
+
+    # Parse the content until the end tag.
+    acceptable_endpartials = ("endpartialdef", f"endpartialdef {partial_name}")
```

Comment:
...but it does highlight that if the tag name is supposed to be dynamic, then this hard-coding of `"endpartialdef"` is wrong and it should be `f"end{tokens[0]}"`.

Further, if it's acceptable to pass in the `partial_name` to the end tag, then we also need to fix `partial_end_tag_re` and perhaps also do some checking in `PartialTemplate.find_partial_source()`? Although the latter would likely be complex, e.g. what if someone has tried to nest `partialdef`s? Is that even supported/valid? I'm guessing from a quick scan of the lines below this it isn't supported. Do we have any tests for that?

Reply:
> Although the latter would likely be complex, e.g. what if someone has tried to nest partialdefs? Is that even supported/valid?

Nested partials **are** supported. This kind of thing is common: 

```html+django
{% extends "base.html" %}

{% block content %}

This is my main page. 

{% partialdef outer inline %}
    It hosts a couple of partials. An outer one... 
    
    (Perhaps a whole list)
    
    {% partialdef inner inine %}
        ... and an inner one. 
        
        (Perhaps just a single list item)
        
    {% endpartialdef inner %}
{% endpartialdef outer %}

{% endblock content %}
```

Reply:
@ngnpope Could you please clarify what do you mean with "it does highlight that if the tag name is supposed to be dynamic, then this hard-coding of "endpartialdef" is wrong and it should be f"end{tokens[0]}"."?